segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/StorageLogManager.java

```diff
@@ -0,0 +1,813 @@
+        }
+        LedgerData ld = new LedgerData(lh, startOffset, stat.getVersion(), containerEpoch);
+        ld.setLength((int) lh.getLength());
+        ld.setReadonly(lh.isClosed());
+        ld.setLastAddConfirmed(lh.getLastAddConfirmed());
+        return ld;
+    }
+
+    private CompletableFuture<StorageLog> getOrRetrieveStorageLedger(String streamSegmentName, boolean readOnly) {
+        CompletableFuture<StorageLog> retVal = new CompletableFuture<>();
+
+        synchronized (this) {
+            if (ledgers.containsKey(streamSegmentName)) {
+                StorageLog ledger = ledgers.get(streamSegmentName);
+                if (!readOnly && ledger.isReadOnlyHandle()) {
+                    //Check whether the value is read-write, if it is not, flush it so that we create a read-write token
+                    ledgers.remove(streamSegmentName);
+                } else {
+                    // If the caller expects a readonly handle, either a readonly or read-write cached value will work.
+                    retVal.complete(ledger);
+                    return retVal;
+                }
+            }
+        }
+        return retrieveStorageLedgerMetadata(streamSegmentName, readOnly);
+    }
+
+    private CompletableFuture<StorageLog> retrieveStorageLedgerMetadata(String streamSegmentName, boolean readOnly) {
+
+        Stat stat = new Stat();
+        return zkClient.getData().storingStatIn(stat).forPath(getZkPath(streamSegmentName)).toCompletableFuture()
+                       .exceptionally(exc -> {
+                           translateZKException(streamSegmentName, exc);
+                           return null;
+                       })
+                       .thenApply(bytes -> {
+                           StorageLog retval = StorageLog.deserialize(this, streamSegmentName, bytes, stat.getVersion());
+                           ledgers.put(streamSegmentName, retval);
+                           return retval;
+                       })
+                       .thenCompose(storageLog -> {
+                           if (readOnly) {
+                               return getBKLEdgerDetails(streamSegmentName, storageLog);
+                           } else {
+                               CompletableFuture<StorageLog> retVal = new CompletableFuture<>();
+                               retVal.complete(storageLog);
+                               return retVal;
+                           }
+                       });
+    }
+
+    private CompletableFuture<StorageLog> getBKLEdgerDetails(String streamSegmentName, StorageLog storageLog) {
+        return zkClient.getChildren().forPath(getZkPath(streamSegmentName))
+                       .toCompletableFuture()
+                       .thenCompose(children -> {
+                           CompletableFuture<LedgerData>[] futures = children.stream()
+                                                                             .map(child -> {
```

Comment:
Can we improve the indentation here? This is pushing too much to the right.

Reply:
Fixed.